File: ios/Classes/ZendeskMessaging.swift
import UIKit
import ZendeskSDKMessaging
import ZendeskSDK
import UserNotifications

public class ZendeskMessaging: NSObject {
    private static let unreadMessages = "unread_messages"
    private static let zendeskEvent = "zendesk_event"

    let TAG = "[ZendeskMessaging]"

    private weak var zendeskPlugin: SwiftZendeskMessagingPlugin?
    private let channel: FlutterMethodChannel
    private var lastConnectionStatus: String = "unknown"

    init(flutterPlugin: SwiftZendeskMessagingPlugin, channel: FlutterMethodChannel) {
        self.zendeskPlugin = flutterPlugin
        self.channel = channel
    }

    func initialize(channelKey: String, flutterResult: @escaping FlutterResult) {
        print("\(self.TAG) - Channel Key - \(channelKey)\n")
        Zendesk.initialize(withChannelKey: channelKey, messagingFactory: DefaultMessagingFactory()) { result in
            DispatchQueue.main.async {
                if case let .failure(error) = result {
                    self.zendeskPlugin?.isInitialized = false
                    print("\(self.TAG) - initialize failure - \(error.localizedDescription)\n")
                    flutterResult(FlutterError(
                        code: "initialize_error",
                        message: error.localizedDescription,
                        details: nil)
                    )
                } else {
                    self.zendeskPlugin?.isInitialized = true
                    print("\(self.TAG) - initialize success")
                    flutterResult(nil)
                }
            }
        }
    }

    func invalidate() {
        Zendesk.instance?.removeEventObserver(self)
        Zendesk.invalidate()
        self.zendeskPlugin?.isInitialized = false
        self.zendeskPlugin?.isLoggedIn = false
        print("\(self.TAG) - invalidate")
    }

    func show(rootViewController: UIViewController?, flutterResult: @escaping FlutterResult) {
        guard let viewController = Zendesk.instance?.messaging?.messagingViewController() else {
            print("\(self.TAG) - Unable to create Zendesk messaging view controller")
            flutterResult(FlutterError(
                code: "show_error",
                message: "Unable to create Zendesk messaging view controller",
                details: nil)
            )
            return
        }
        guard let rootViewController = rootViewController else {
            print("\(self.TAG) - Root view controller is nil")
            flutterResult(FlutterError(
                code: "show_error",
                message: "Root view controller is nil",
                details: nil)
            )
            return
        }

        let navController = UINavigationController(rootViewController: viewController)

        DispatchQueue.main.async {
            if let presentedVC = rootViewController.presentedViewController {
                if presentedVC !== navController {
                    presentedVC.dismiss(animated: true) {
                        rootViewController.present(navController, animated: true, completion: nil)
                    }
                } else {
                    print("\(self.TAG) - Zendesk messaging view controller is already presented")
                }
            } else {
                rootViewController.present(navController, animated: true, completion: nil)
            }
            flutterResult(nil)
        }
        print("\(self.TAG) - show")
    }

    func showConversation(conversationId: String, rootViewController: UIViewController?, flutterResult: @escaping FlutterResult) {
        guard let viewController = Zendesk.instance?.messaging?.messagingViewController(
            .showConversation(conversationId: conversationId, exitAction: .returnToConversationList)
        ) else {
            print("\(self.TAG) - Unable to create Zendesk messaging view controller for conversation")
            flutterResult(FlutterError(
                code: "show_error",
                message: "Unable to create Zendesk messaging view controller for conversation",
                details: nil)
            )
            return
        }
        guard let rootViewController = rootViewController else {
            print("\(self.TAG) - Root view controller is nil")
            flutterResult(FlutterError(
                code: "show_error",
                message: "Root view controller is nil",
                details: nil)
            )
            return
        }

        let navController = UINavigationController(rootViewController: viewController)

        DispatchQueue.main.async {
            if let presentedVC = rootViewController.presentedViewController {
                presentedVC.dismiss(animated: true) {
                    rootViewController.present(navController, animated: true, completion: nil)
                }
            } else {
                rootViewController.present(navController, animated: true, completion: nil)
            }
            flutterResult(nil)
        }
        print("\(self.TAG) - showConversation: \(conversationId)")
    }

    func showConversationList(rootViewController: UIViewController?, flutterResult: @escaping FlutterResult) {
        guard let viewController = Zendesk.instance?.messaging?.messagingViewController(
            .showConversationList
        ) else {
            print("\(self.TAG) - Unable to create Zendesk conversation list view controller")
            flutterResult(FlutterError(
                code: "show_error",
                message: "Unable to create Zendesk conversation list view controller",
                details: nil)
            )
            return
        }
        guard let rootViewController = rootViewController else {
            print("\(self.TAG) - Root view controller is nil")
            flutterResult(FlutterError(
                code: "show_error",
                message: "Root view controller is nil",
                details: nil)
            )
            return
        }

        let navController = UINavigationController(rootViewController: viewController)

        DispatchQueue.main.async {
            if let presentedVC = rootViewController.presentedViewController {
                presentedVC.dismiss(animated: true) {
                    rootViewController.present(navController, animated: true, completion: nil)
                }
            } else {
                rootViewController.present(navController, animated: true, completion: nil)
            }
            flutterResult(nil)
        }
        print("\(self.TAG) - showConversationList")
    }

    func startNewConversation(rootViewController: UIViewController?, flutterResult: @escaping FlutterResult) {
        guard let viewController = Zendesk.instance?.messaging?.messagingViewController(
            .showNewConversation(exitAction: .returnToConversationList)
        ) else {
            print("\(self.TAG) - Unable to create Zendesk new conversation view controller")
            flutterResult(FlutterError(
                code: "show_error",
                message: "Unable to create Zendesk new conversation view controller",
                details: nil)
            )
            return
        }
        guard let rootViewController = rootViewController else {
            print("\(self.TAG) - Root view controller is nil")
            flutterResult(FlutterError(
                code: "show_error",
                message: "Root view controller is nil",
                details: nil)
            )
            return
        }

        let navController = UINavigationController(rootViewController: viewController)

        DispatchQueue.main.async {
            if let presentedVC = rootViewController.presentedViewController {
                presentedVC.dismiss(animated: true) {
                    rootViewController.present(navController, animated: true, completion: nil)
                }
            } else {
                rootViewController.present(navController, animated: true, completion: nil)
            }
            flutterResult(nil)
        }
        print("\(self.TAG) - startNewConversation")
    }

    func setConversationTags(tags: [String]) {
        Zendesk.instance?.messaging?.setConversationTags(tags)
        print("\(self.TAG) - setConversationTags: \(tags)")
    }

    func clearConversationTags() {
        Zendesk.instance?.messaging?.clearConversationTags()
        print("\(self.TAG) - clearConversationTags")
    }

    func loginUser(jwt: String, flutterResult: @escaping FlutterResult) {
        Zendesk.instance?.loginUser(with: jwt) { result in
            DispatchQueue.main.async {
                switch result {
                case .success(let user):
                    self.zendeskPlugin?.isLoggedIn = true
                    flutterResult([
                        "id": user.id,
                        "externalId": user.externalId,
                        "authenticationType": self.getAuthenticationType(user: user)
                    ])
                    print("\(self.TAG) - loginUser success")
                case .failure(let error):
                    print("\(self.TAG) - login failure - \(error.localizedDescription)\n")
                    flutterResult(FlutterError(
                        code: "login_error",
                        message: error.localizedDescription,
                        details: nil)
                    )
                }
            }
        }
    }

    func logoutUser(flutterResult: @escaping FlutterResult) {
        Zendesk.instance?.logoutUser { result in
            DispatchQueue.main.async {
                switch result {
                case .success:
                    self.zendeskPlugin?.isLoggedIn = false
                    flutterResult(nil)
                    print("\(self.TAG) - logoutUser success")
                case .failure(let error):
                    print("\(self.TAG) - logout failure - \(error.localizedDescription)\n")
                    flutterResult(FlutterError(
                        code: "logout_error",
                        message: error.localizedDescription,
                        details: nil)
                    )
                }
            }
        }
    }

    func getCurrentUser(flutterResult: @escaping FlutterResult) {
        if let user = Zendesk.instance?.getCurrentUser() {
            flutterResult([
                "id": user.id,
                "externalId": user.externalId,
                "authenticationType": getAuthenticationType(user: user)
            ])
        } else {
            flutterResult(nil)
        }
    }

    private func getAuthenticationType(user: ZendeskSDK.ZendeskUser) -> String {
        switch user.authenticationType {
        case .jwt:
            return "jwt"
        default:
            return "anonymous"
        }
    }

    func getUnreadMessageCount() -> Int {
        let count = Zendesk.instance?.messaging?.getUnreadMessageCount()
        return count ?? 0
    }

    func getUnreadMessageCountForConversation(conversationId: String) -> Int {
        let count = Zendesk.instance?.messaging?.getUnreadMessageCount(conversationId: conversationId)
        return count ?? 0
    }

    func getConnectionStatus() -> String {
        // Connection status is obtained from events
        return "unknown"
    }

    func listenMessageCountChanged() {
        Zendesk.instance?.addEventObserver(self, { event in
            self.handleZendeskEvent(event: event)
        })
        print("\(self.TAG) - listenMessageCountChanged - Event observer added")
    }

    private func handleZendeskEvent(event: ZendeskSDK.ZendeskEvent) {
        switch event {
        case let .unreadMessageCountChanged(currentUnreadCount):
            // Legacy callback for backwards compatibility
            self.channel.invokeMethod(
                Self.unreadMessages,
                arguments: ["messages_count": currentUnreadCount]
            )
            // New event system
            self.channel.invokeMethod(
                Self.zendeskEvent,
                arguments: [
                    "type": "unreadMessageCountChanged",
                    "timestamp": Int64(Date().timeIntervalSince1970 * 1000),
                    "totalUnreadCount": currentUnreadCount
                ]
            )

        case let .authenticationFailed(error as NSError):
            let isJwtExpired = error.localizedDescription.lowercased().contains("expired")
            self.channel.invokeMethod(
                Self.zendeskEvent,
                arguments: [
                    "type": "authenticationFailed",
                    "timestamp": Int64(Date().timeIntervalSince1970 * 1000),
                    "errorCode": "authentication_failed",
                    "errorMessage": error.localizedDescription,
                    "isJwtExpired": isJwtExpired
                ]
            )

        case let .conversationAdded(conversationId):
            self.channel.invokeMethod(
                Self.zendeskEvent,
                arguments: [
                    "type": "conversationAdded",
                    "timestamp": Int64(Date().timeIntervalSince1970 * 1000),
                    "conversationId": conversationId
                ]
            )

        case let .connectionStatusChanged(connectionStatus):
            let statusString = connectionStatus.stringValue
            self.channel.invokeMethod(
                Self.zendeskEvent,
                arguments: [
                    "type": "connectionStatusChanged",
                    "timestamp": Int64(Date().timeIntervalSince1970 * 1000),
                    "status": statusString
                ]
            )

        case let .sendMessageFailed(error as NSError):
            self.channel.invokeMethod(
                Self.zendeskEvent,
                arguments: [
                    "type": "sendMessageFailed",
                    "timestamp": Int64(Date().timeIntervalSince1970 * 1000),
                    "errorMessage": error.localizedDescription
                ]
            )

        case let .conversationOpened(id, timestamp, conversationId):
            self.channel.invokeMethod(
                Self.zendeskEvent,
                arguments: [
                    "type": "conversationOpened",
                    "id": id.uuidString,
                    "timestamp": Int64(timestamp.timeIntervalSince1970 * 1000),
                    "conversationId": conversationId ?? ""
                ]
            )

        case let .conversationStarted(id, timestamp, conversationId):
            self.channel.invokeMethod(
                Self.zendeskEvent,
                arguments: [
                    "type": "conversationStarted",
                    "id": id.uuidString,
                    "timestamp": Int64(timestamp.timeIntervalSince1970 * 1000),
                    "conversationId": conversationId
                ]
            )

        case let .messagesShown(id, timestamp, conversationId, messages):
            let messagesData = messages.map { message -> [String: Any] in
                return [
                    "id": message.id,
                    "conversationId": conversationId
                ]
            }
            self.channel.invokeMethod(
                Self.zendeskEvent,
                arguments: [
                    "type": "messagesShown",
                    "id": id.uuidString,
                    "timestamp": Int64(timestamp.timeIntervalSince1970 * 1000),
                    "conversationId": conversationId,
                    "messages": messagesData
                ]
            )

        case let .сonversationUnreadCountChanged(id, timestamp, data):
            self.channel.invokeMethod(
                Self.zendeskEvent,
                arguments: [
                    "type": "unreadMessageCountChanged",
                    "id": id.uuidString,
                    "timestamp": Int64(timestamp.timeIntervalSince1970 * 1000),
                    "conversationId": data.conversationId ?? "",
                    "conversationUnreadCount": data.unreadCountInConversation,
                    "totalUnreadCount": data.totalUnreadMessagesCount
                ]
            )

        @unknown default:
            print("\(self.TAG) - Unknown event type")
        }
    }

    func setConversationFields(fields: [String: String]) {
        Zendesk.instance?.messaging?.setConversationFields(fields)
        print("\(self.TAG) - setConversationFields: \(fields)")
    }

    func clearConversationFields() {
        Zendesk.instance?.messaging?.clearConversationFields()
        print("\(self.TAG) - clearConversationFields")
    }

    // ============================================================================
    // Push Notifications
    // ============================================================================

    /// Update the push notification token with Zendesk.
    /// Call this when receiving a new APNs device token.
    func updatePushNotificationToken(_ deviceToken: Data) {
        PushNotifications.updatePushNotificationToken(deviceToken)
        print("\(self.TAG) - updatePushNotificationToken: token updated")
    }

    /// Update the push notification token from a string (FCM token format).
    /// Converts the string to Data before passing to SDK.
    func updatePushNotificationTokenString(_ token: String) {
        // For iOS, we typically receive Data from APNs, but if using FCM,
        // the token comes as a string. We pass it directly to the SDK.
        if let tokenData = token.data(using: .utf8) {
            PushNotifications.updatePushNotificationToken(tokenData)
            print("\(self.TAG) - updatePushNotificationTokenString: token updated")
        } else {
            print("\(self.TAG) - updatePushNotificationTokenString: invalid token format")
        }
    }

    /// Check if a push notification should be displayed by Zendesk.
    /// Returns the responsibility indicating how to handle the notification.
    func shouldBeDisplayed(_ userInfo: [AnyHashable: Any]) -> String {
        let responsibility = PushNotifications.shouldBeDisplayed(userInfo)
        let result: String
        switch responsibility {
        case .messagingShouldDisplay:
            result = "messaging_should_display"
        case .messagingShouldNotDisplay:
            result = "messaging_should_not_display"
        case .notFromMessaging:
            result = "not_from_messaging"
        @unknown default:
            result = "unknown"
        }
        print("\(self.TAG) - shouldBeDisplayed: \(result)")
        return result
    }

    /// Handle and display a push notification.
    /// Returns true if the notification was handled by Zendesk.
    func handleNotification(_ userInfo: [AnyHashable: Any]) -> Bool {
        let responsibility = PushNotifications.shouldBeDisplayed(userInfo)
        if responsibility == .messagingShouldDisplay {
            print("\(self.TAG) - handleNotification: Zendesk notification detected")
            return true
        } else {
            print("\(self.TAG) - handleNotification: not a Zendesk notification")
            return false
        }
    }

    /// Handle a notification tap event.
    /// Returns the view controller to display, or nil if not a Zendesk notification.
    func handleNotificationTap(_ userInfo: [AnyHashable: Any], rootViewController: UIViewController?, completion: @escaping (Bool) -> Void) {
        let responsibility = PushNotifications.shouldBeDisplayed(userInfo)
        if responsibility == .messagingShouldDisplay {
            PushNotifications.handleTap(userInfo) { [weak self] viewController in
                guard let self = self else {
                    completion(false)
                    return
                }
                if let vc = viewController, let rootVC = rootViewController {
                    let navController = UINavigationController(rootViewController: vc)
                    DispatchQueue.main.async {
                        if let presentedVC = rootVC.presentedViewController {
                            presentedVC.dismiss(animated: true) {
                                rootVC.present(navController, animated: true, completion: nil)
                            }
                        } else {
                            rootVC.present(navController, animated: true, completion: nil)
                        }
                        completion(true)
                    }
                    print("\(self.TAG) - handleNotificationTap: opened messaging")
                } else {
                    print("\(self.TAG) - handleNotificationTap: viewController is nil (app may have been killed)")
                    completion(false)
                }
            }
        } else {
            print("\(self.TAG) - handleNotificationTap: not a Zendesk notification")
            completion(false)
        }
    }
}

File: android/src/main/kotlin/com/chyiiiiiiiiiiiiii/zendesk_messaging/ZendeskMessaging.kt
package com.chyiiiiiiiiiiiiii.zendesk_messaging

import android.content.Context
import io.flutter.plugin.common.MethodChannel
import zendesk.android.Zendesk
import zendesk.android.ZendeskUser
import zendesk.android.events.ZendeskEvent
import zendesk.android.events.ZendeskEventListener
import zendesk.android.messaging.MessagingScreen
import zendesk.messaging.android.DefaultMessagingFactory
import zendesk.messaging.android.push.PushNotifications
import zendesk.messaging.android.push.PushResponsibility

class ZendeskMessaging(
    private val plugin: ZendeskMessagingPlugin,
    private val channel: MethodChannel
) {
    companion object {
        const val TAG = "[ZendeskMessaging]"

        // Method channel event keys
        const val EVENT_UNREAD_MESSAGES = "unread_messages"
        const val EVENT_ZENDESK_EVENT = "zendesk_event"
    }

    // Event listener for all Zendesk events
    private val zendeskEventListener = ZendeskEventListener { zendeskEvent ->
        handleZendeskEvent(zendeskEvent)
    }

    private fun handleZendeskEvent(zendeskEvent: ZendeskEvent) {
        when (zendeskEvent) {
            is ZendeskEvent.UnreadMessageCountChanged -> {
                // Legacy callback for backwards compatibility
                channel.invokeMethod(
                    EVENT_UNREAD_MESSAGES,
                    mapOf("messages_count" to zendeskEvent.currentUnreadCount)
                )
                // New event system
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "unreadMessageCountChanged",
                        "timestamp" to System.currentTimeMillis(),
                        "totalUnreadCount" to zendeskEvent.currentUnreadCount
                    )
                )
            }

            is ZendeskEvent.AuthenticationFailed -> {
                val isJwtExpired = zendeskEvent.error.message?.contains("expired", ignoreCase = true) == true
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "authenticationFailed",
                        "timestamp" to System.currentTimeMillis(),
                        "errorCode" to "authentication_failed",
                        "errorMessage" to (zendeskEvent.error.message ?: "Unknown error"),
                        "isJwtExpired" to isJwtExpired
                    )
                )
            }

            is ZendeskEvent.FieldValidationFailed -> {
                val errorMessages = zendeskEvent.errors.map { it.toString() }
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "fieldValidationFailed",
                        "timestamp" to System.currentTimeMillis(),
                        "errors" to errorMessages
                    )
                )
            }

            is ZendeskEvent.ConnectionStatusChanged -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "connectionStatusChanged",
                        "timestamp" to System.currentTimeMillis(),
                        "status" to zendeskEvent.connectionStatus.name.lowercase()
                    )
                )
            }

            is ZendeskEvent.SendMessageFailed -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "sendMessageFailed",
                        "timestamp" to System.currentTimeMillis(),
                        "errorMessage" to (zendeskEvent.cause.message ?: "Unknown error")
                    )
                )
            }

            is ZendeskEvent.ConversationAdded -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "conversationAdded",
                        "timestamp" to System.currentTimeMillis(),
                        "conversationId" to zendeskEvent.conversationId
                    )
                )
            }

            is ZendeskEvent.ConversationStarted -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "conversationStarted",
                        "timestamp" to System.currentTimeMillis(),
                        "conversationId" to zendeskEvent.conversationId
                    )
                )
            }

            is ZendeskEvent.ConversationOpened -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "conversationOpened",
                        "timestamp" to System.currentTimeMillis(),
                        "conversationId" to zendeskEvent.conversationId
                    )
                )
            }

            is ZendeskEvent.MessagesShown -> {
                val messagesData = zendeskEvent.messages.map { message ->
                    mapOf(
                        "id" to message.id,
                        "conversationId" to zendeskEvent.conversationId
                    )
                }
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "messagesShown",
                        "timestamp" to System.currentTimeMillis(),
                        "conversationId" to zendeskEvent.conversationId,
                        "messages" to messagesData
                    )
                )
            }

            is ZendeskEvent.ProactiveMessageDisplayed -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "proactiveMessageDisplayed",
                        "timestamp" to System.currentTimeMillis(),
                        "proactiveMessageId" to "",
                        "campaignId" to null
                    )
                )
            }

            is ZendeskEvent.ProactiveMessageClicked -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "proactiveMessageClicked",
                        "timestamp" to System.currentTimeMillis(),
                        "proactiveMessageId" to "",
                        "campaignId" to null
                    )
                )
            }

            is ZendeskEvent.ConversationWithAgentRequested -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "conversationWithAgentRequested",
                        "timestamp" to System.currentTimeMillis(),
                        "conversationId" to ""
                    )
                )
            }

            is ZendeskEvent.ConversationServedByAgent -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "conversationServedByAgent",
                        "timestamp" to System.currentTimeMillis(),
                        "conversationId" to ""
                    )
                )
            }

            is ZendeskEvent.MessagingOpened -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "messagingOpened",
                        "timestamp" to System.currentTimeMillis()
                    )
                )
            }

            is ZendeskEvent.MessagingClosed -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "messagingClosed",
                        "timestamp" to System.currentTimeMillis()
                    )
                )
            }

            is ZendeskEvent.NewConversationButtonClicked -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "newConversationButtonClicked",
                        "timestamp" to System.currentTimeMillis()
                    )
                )
            }

            is ZendeskEvent.PostbackButtonClicked -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "postbackButtonClicked",
                        "timestamp" to System.currentTimeMillis(),
                        "conversationId" to "",
                        "actionName" to ""
                    )
                )
            }

            is ZendeskEvent.ConversationExtensionOpened -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "conversationExtensionOpened",
                        "timestamp" to System.currentTimeMillis(),
                        "conversationId" to "",
                        "extensionUrl" to ""
                    )
                )
            }

            is ZendeskEvent.ConversationExtensionDisplayed -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "conversationExtensionDisplayed",
                        "timestamp" to System.currentTimeMillis(),
                        "conversationId" to "",
                        "extensionUrl" to ""
                    )
                )
            }

            is ZendeskEvent.ArticleBrowserClicked -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "articleBrowserClicked",
                        "timestamp" to System.currentTimeMillis(),
                        "articleUrl" to ""
                    )
                )
            }

            is ZendeskEvent.ArticleClicked -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "articleClicked",
                        "timestamp" to System.currentTimeMillis(),
                        "articleUrl" to "",
                        "conversationId" to ""
                    )
                )
            }

            is ZendeskEvent.NotificationDisplayed -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "notificationDisplayed",
                        "timestamp" to System.currentTimeMillis(),
                        "conversationId" to ""
                    )
                )
            }

            is ZendeskEvent.NotificationOpened -> {
                channel.invokeMethod(
                    EVENT_ZENDESK_EVENT,
                    mapOf(
                        "type" to "notificationOpened",
                        "timestamp" to System.currentTimeMillis(),
                        "conversationId" to ""
                    )
                )
            }

            else -> {
                // Default branch for forward compatibility with Zendesk SDK and its `ZendeskEvent` expansion
                println("$TAG - Unknown event type: $zendeskEvent")
            }
        }
    }

    fun initialize(channelKey: String, result: MethodChannel.Result) {
        println("$TAG - Channel Key - $channelKey")
        Zendesk.initialize(
            plugin.activity!!,
            channelKey,
            successCallback = { value ->
                plugin.isInitialized = true
                println("$TAG - initialize success - $value")
                result.success(null)
            },
            failureCallback = { error ->
                plugin.isInitialized = false
                println("$TAG - initialize failure - $error")
                result.error("initialize_error", error.message, null)
            },
            messagingFactory = DefaultMessagingFactory()
        )
    }

    fun invalidate() {
        Zendesk.instance.removeEventListener(zendeskEventListener)
        Zendesk.invalidate()
        plugin.isInitialized = false
        plugin.isLoggedIn = false
        println("$TAG - invalidated")
    }

    fun show() {
        Zendesk.instance.messaging.showMessaging(
            plugin.activity!!,
            MessagingScreen.MostRecentActiveConversation()
        )
        println("$TAG - show")
    }

    fun showConversation(conversationId: String) {
        Zendesk.instance.messaging.showMessaging(
            plugin.activity!!,
            MessagingScreen.Conversation(id = conversationId)
        )
        println("$TAG - showConversation: $conversationId")
    }

    fun showConversationList() {
        Zendesk.instance.messaging.showMessaging(
            plugin.activity!!,
            MessagingScreen.ConversationsList
        )
        println("$TAG - showConversationList")
    }

    fun startNewConversation() {
        Zendesk.instance.messaging.showMessaging(
            plugin.activity!!,
            MessagingScreen.NewConversation()
        )
        println("$TAG - startNewConversation")
    }

    fun getUnreadMessageCount(): Int =
        try {
            Zendesk.instance.messaging.getUnreadMessageCount()
        } catch (error: Throwable) {
            println("$TAG - getUnreadMessageCount error: ${error.message}")
            0
        }

    fun getUnreadMessageCountForConversation(conversationId: String): Int =
        try {
            Zendesk.instance.messaging.getUnreadMessageCount(conversationId)
        } catch (error: Throwable) {
            println("$TAG - getUnreadMessageCountForConversation error: ${error.message}")
            0
        }

    fun setConversationTags(tags: List<String>) {
        Zendesk.instance.messaging.setConversationTags(tags)
        println("$TAG - setConversationTags: $tags")
    }

    fun clearConversationTags() {
        Zendesk.instance.messaging.clearConversationTags()
        println("$TAG - clearConversationTags")
    }

    fun loginUser(jwt: String, result: MethodChannel.Result) {
        Zendesk.instance.loginUser(
            jwt,
            { user ->
                plugin.isLoggedIn = true
                result.success(
                    mapOf(
                        "id" to user.id,
                        "externalId" to user.externalId,
                        "authenticationType" to getAuthenticationType(user)
                    )
                )
                println("$TAG - loginUser success")
            },
            { error ->
                println("$TAG - Login failure : ${error.message}")
                result.error("login_error", error.message, null)
            }
        )
    }

    fun logoutUser(result: MethodChannel.Result) {
        Zendesk.instance.logoutUser(
            successCallback = {
                plugin.isLoggedIn = false
                result.success(null)
                println("$TAG - logoutUser success")
            },
            failureCallback = { error ->
                println("$TAG - Logout failure : ${error.message}")
                result.error("logout_error", error.message, null)
            }
        )
        Zendesk.instance.removeEventListener(zendeskEventListener)
    }

    fun getCurrentUser(result: MethodChannel.Result) {
        try {
            Zendesk.instance.getCurrentUser { user ->
                if (user != null) {
                    result.success(
                        mapOf(
                            "id" to user.id,
                            "externalId" to user.externalId,
                            "authenticationType" to getAuthenticationType(user)
                        )
                    )
                } else {
                    result.success(null)
                }
            }
        } catch (error: Throwable) {
            println("$TAG - getCurrentUser error: ${error.message}")
            result.success(null)
        }
    }

    private fun getAuthenticationType(user: ZendeskUser): String {
        return try {
            when (user.authenticationType) {
                zendesk.android.ZendeskAuthenticationType.Jwt -> "jwt"
                else -> "anonymous"
            }
        } catch (e: Throwable) {
            "anonymous"
        }
    }

    fun getConnectionStatus(): String {
        return try {
            // Connection status is obtained from events, return current known state
            "unknown"
        } catch (error: Throwable) {
            println("$TAG - getConnectionStatus error: ${error.message}")
            "unknown"
        }
    }

    fun listenMessageCountChanged() {
        // To add the event listener to your Zendesk instance:
        Zendesk.instance.addEventListener(zendeskEventListener)
        println("$TAG - listenMessageCountChanged - Event listener added")
    }

    fun setConversationFields(fields: Map<String, String>) {
        Zendesk.instance.messaging.setConversationFields(fields)
        println("$TAG - setConversationFields: $fields")
    }

    fun clearConversationFields() {
        Zendesk.instance.messaging.clearConversationFields()
        println("$TAG - clearConversationFields")
    }

    // ============================================================================
    // Push Notifications
    // ============================================================================

    /**
     * Update the push notification token with Zendesk.
     * Call this when receiving a new FCM token.
     */
    fun updatePushNotificationToken(token: String) {
        try {
            PushNotifications.updatePushNotificationToken(token)
            println("$TAG - updatePushNotificationToken: token updated")
        } catch (error: Throwable) {
            println("$TAG - updatePushNotificationToken error: ${error.message}")
            throw error
        }
    }

    /**
     * Check if a push notification should be displayed by Zendesk.
     * Returns the responsibility indicating how to handle the notification.
     */
    fun shouldBeDisplayed(messageData: Map<String, String>): String {
        return try {
            val responsibility = PushNotifications.shouldBeDisplayed(messageData)
            val result = when (responsibility) {
                PushResponsibility.MESSAGING_SHOULD_DISPLAY -> "messaging_should_display"
                PushResponsibility.MESSAGING_SHOULD_NOT_DISPLAY -> "messaging_should_not_display"
                PushResponsibility.NOT_FROM_MESSAGING -> "not_from_messaging"
                else -> "unknown"
            }
            println("$TAG - shouldBeDisplayed: $result")
            result
        } catch (error: Throwable) {
            println("$TAG - shouldBeDisplayed error: ${error.message}")
            "unknown"
        }
    }

    /**
     * Handle and display a push notification.
     * Returns true if the notification was handled by Zendesk.
     */
    fun handleNotification(context: Context, messageData: Map<String, String>): Boolean {
        return try {
            val responsibility = PushNotifications.shouldBeDisplayed(messageData)
            if (responsibility == PushResponsibility.MESSAGING_SHOULD_DISPLAY) {
                PushNotifications.displayNotification(context, messageData)
                println("$TAG - handleNotification: notification displayed")
                true
            } else {
                println("$TAG - handleNotification: not a Zendesk notification")
                false
            }
        } catch (error: Throwable) {
            println("$TAG - handleNotification error: ${error.message}")
            false
        }
    }

    /**
     * Handle a notification tap event.
     * Opens the messaging UI to the relevant conversation.
     */
    fun handleNotificationTap(context: Context, messageData: Map<String, String>) {
        try {
            val responsibility = PushNotifications.shouldBeDisplayed(messageData)
            if (responsibility == PushResponsibility.MESSAGING_SHOULD_DISPLAY) {
                // Show messaging UI - the SDK will navigate to the correct conversation
                Zendesk.instance.messaging.showMessaging(
                    plugin.activity!!,
                    MessagingScreen.MostRecentActiveConversation()
                )
                println("$TAG - handleNotificationTap: opened messaging")
            } else {
                println("$TAG - handleNotificationTap: not a Zendesk notification")
            }
        } catch (error: Throwable) {
            println("$TAG - handleNotificationTap error: ${error.message}")
            throw error
        }
    }
}
